nova/compute/api.py

```diff
@@ -1773,10 +1774,15 @@ def soft_delete(self, context, instance):
         """Terminate an instance."""
         LOG.debug('Going to try to soft delete instance',
                   instance=instance)
-
-        self._delete(context, instance, 'soft_delete', self._do_soft_delete,
-                     task_state=task_states.SOFT_DELETING,
-                     deleted_at=timeutils.utcnow())
+        if instance['vm_state'] == vm_states.SOFT_DELETED:
+            LOG.warn("Instance %s is soft delete state" % (instance.uuid))
+            reason = ("The instance is %s state" % (instance.vm_state))
+            raise exception.InstanceTerminationFailure(reason=reason)
+        else:
+            self._delete(context, instance, 'soft_delete',
+                         self._do_soft_delete,
+                         task_state=task_states.SOFT_DELETING,
+                         deleted_at=timeutils.utcnow())
 
```

Comment:
这个地方引入异常抛出目的在于告诉测试以及后台使用人员，当一个处于已经删除状态的云主机，再次进行删除时会提示错误，如果不提示错误，测试或者后台使用人员会觉得一直进行删除，但都没有报不存在或其它错误，会感觉这是一个bug，怎么删除都没有问题的bug
．

2016-07-08 13:35 GMT+08:00 daemon <notifications@github.com>:

> In nova/compute/api.py
> <https://github.com/eayunstack/nova/pull/37#discussion_r70027047>:
>
> > @@ -1773,10 +1774,15 @@ def soft_delete(self, context, instance):
> >          """Terminate an instance."""
> >          LOG.debug('Going to try to soft delete instance',
> >                    instance=instance)
> > -
> > -        self._delete(context, instance, 'soft_delete', self._do_soft_delete,
> > -                     task_state=task_states.SOFT_DELETING,
> > -                     deleted_at=timeutils.utcnow())
> > +        if instance['vm_state'] == vm_states.SOFT_DELETED:
> > +            LOG.warn("Instance %s is soft delete state" % (instance.uuid))
> > +            reason = ("The instance is %s state" % (instance.vm_state))
> > +            raise exception.InstanceTerminationFailure(reason=reason)
> > +        else:
> > +            self._delete(context, instance, 'soft_delete',
> > +                         self._do_soft_delete,
> > +                         task_state=task_states.SOFT_DELETING,
> > +                         deleted_at=timeutils.utcnow())
> >
>
> 我不反对打印提示信息，我只是建议在那基础上修改，当然这样也没什么问题
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/eayunstack/nova/pull/37/files/b9f59ddec6dfb4785ef13559f31c37d1239b21e7#r70027047>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AKMiZkO1CCGquwcwwzbYweDkk-86i13rks5qTeGFgaJpZM4I0Frh>
> .
>